tools/resourcedocsgen/cmd/docs/registry.go

```diff
@@ -126,31 +129,71 @@ func genResourceDocsForAllRegistryPackages(registryRepoPath, baseDocsOutDir, bas
 		return errors.Wrap(err, "reading the registry packages dir")
 	}
 
-	for _, f := range metadataFiles {
-		glog.Infof("=== starting %s ===\n", f.Name())
-		glog.Infoln("Processing metadata file")
-		metadataFilePath := filepath.Join(registryPackagesPath, f.Name())
+	var m sync.Mutex
+	var errs []error
+	addError := func(err error) {
+		m.Lock()
+		defer m.Unlock()
+		errs = append(errs, err)
+	}
 
-		b, err := os.ReadFile(metadataFilePath)
-		if err != nil {
-			return errors.Wrapf(err, "reading the metadata file %s", metadataFilePath)
-		}
+	// Concurrency is limited via a checkout-checkin system.
+	//
+	// Each time a process starts, it checks out a hallpass from the monitor (blocking
+	// if none are available). When the process completes, it returns it's hallpass to
+	// the monitor.
+	//
+	// Because there are a fixed number of hallpasses (maxConcurrency), and each
+	// process needs a hallpass to run, we limit effective concurrency to the number
+	// of available hallpasses.
+	maxConcurrency := runtime.NumCPU()
+
+	var hallpass struct{}
+	monitor := make(chan struct{}, maxConcurrency)
+	for i := 0; i < maxConcurrency; i++ {
+		monitor <- hallpass
+	}
 
-		var metadata pkg.PackageMeta
-		if err := yaml.Unmarshal(b, &metadata); err != nil {
-			return errors.Wrapf(err, "unmarshalling the metadata file %s", metadataFilePath)
-		}
+	var w sync.WaitGroup
+	w.Add(len(metadataFiles))
+	for _, f := range metadataFiles {
+		go func(f os.DirEntry) {
+			defer func(hallpass struct{}) {
+				w.Done()
+				monitor <- hallpass
+			}(<-monitor)
+			glog.Infof("=== starting %s ===\n", f.Name())
+			glog.Infoln("Processing metadata file")
+			metadataFilePath := filepath.Join(registryPackagesPath, f.Name())
+
+			b, err := os.ReadFile(metadataFilePath)
+			if err != nil {
+				addError(errors.Wrapf(err, "reading the metadata file %s", metadataFilePath))
+				return
+			}
```

Comment:
Would be nice to not have the core process of file operations etc intermingled with the concurrency concerns. Perhaps some refactoring to pull out a single method that the concurrency-related code can call into would suffice.